### PR TITLE
New podman::quadlet type for quadlet production

### LINKED ISSUE #76: RFC - Quadlet Support
podman 4.4 has quadlet support  

Is it realistic to just have puppet create quadlet files and then let   the generator create the units before puppet starts them.

### PR DESCRIPTION
**Pull request by traylenator**
Assuming at least podman 4.4.0 the new type `podman::quadlet` will create a quadlet unit file for a container, volume or pod which will in turn generate a service unit for managing the resulting container, volume or pod.

Note that pod quadlets require podman at least version 5.

* Fixes #76

Testing with a simple puppet file:

```puppet
# test.pp
podman::quadlet{'centos.container':
  ensure          => present,
  unit_entry     => {
   'Description' => 'Trivial Container that will be very lazy',
  },
  service_entry       => {
    'TimeoutStartSec' => '900',
  },
  container_entry => {
    'Image' => 'quay.io/centos/centos:latest',
    'Exec'  => 'sh -c "sleep inf'
  },
  install_entry   => {
    'WantedBy' => 'default.target'
  },
  active          => true,
}

```

Results in a running container.

### CODE REVIEW

**Comment by breca** [comment]
This looks great! Would love to play with this.

Not sure if it's intentional but some of your files/classes are named `quadret` 

i.e `types/quadret/container.pp` / `Podman::Quadret::Container`

**Comment by traylenator** [comment]
> Not sure if it's intentional but some of your files/classes are named `quadret`
> 
> i.e `types/quadret/container.pp` / `Podman::Quadret::Container`

That's indeed a  typo from me - will correct.

**Comment by traylenator** [comment]
Been looking at this myself a bit today - I'd probably do it different than this 

`podman::quadlet` is much more sensible choice - they are very much all the same.

**Comment by southalc** [comment]
Thanks for your work and contribution to the project.  I'm sorry about being slow to respond.  I'm not as active here since changing jobs.
This is an interesting capability and can really simplify this module.  I was reviewing the docs and saw that podman quadlet supports container, volume, network, kube, image, and pod.  I like the idea of implementing this as a more generic defined type "podman::quadlet" that supports resources of all these types.  Are you already working on this per your comment the other day?

**Comment by traylenator** [comment]
I do indeed have a branch I'm playing with - more to do I'll update this one or provide a new one later this week.

> I'm not as active here since changing jobs.

voxpupuli ?

**Comment by traylenator** [comment]
@southalc everything stuffed into podman::quadlet now. Only container, volume and pod done but others easy enough to do for someone.

**Comment by traylenator** [comment]
Hi.

A ping basically - voxpupuli ? There's a load of new stuff in the systemd module just released and we could rework a lot of whats here and make this module smaller  - all the handling of user units is now in that module.

File: README.md [hunk 1]
```diff
@@ -189,6 +189,36 @@ Several additional examples are in a separate [github project](https://github.co
...
 
+## Creating Containers,Volumes, Pods, ... with Quadlet Unit Files
+
+Container unit files can be created and managed, for example:
+
+```puppet
+podman::quadlet{'centos.container':
+  ensure          => present,
+  unit_entry     => {
+   'Description' => 'Trivial Container that will be very lazy',
```
Comment (evgeni) [inline README.md:200]:
```suggestion
    'Description' => 'Trivial Container that will be very lazy',
```

(puppet-lint complains if you copy paste the example XD)

File: types/unit/container.pp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  Optional['LogDriver'] => Enum['k8s-file','journald','none','passthrough'],
+  Optional['Mask'] => String[1],
+  Optional['Mount'] => Array[String[1],0],
+  Optional['Network'] => String[1],
+  Optional['NoNewPrivileges'] => Boolean,
+  Optional['Notify'] => Boolean,
+  Optional['PidsLimits'] => Integer[-1],
+  Optional['Pod'] => Pattern[/^[a-zA-Z0-0_-]+\.pod$/],
+  Optional['PodmanArgs'] => Array[String[1],0],
+  Optional['PublishPort'] => Array[Stdlib::Port,1],
```
[the remark this answers is not in the document]
Comment (traylenator) [reply types/unit/container.pp:39]: Do we try anything more complicated that a string ? ... Maybe `Array[Variant[String[1],Stdlib::Port,1]` so at least a port is accepted - a common use case.

[the remark this answers is not in the document]
Comment (ekohl) [reply types/unit/container.pp]: Sounds like a good idea to me. You can always expand it later to a `Pattern`.

**Comment by southalc** [comment]
I finally got around to this, although I see the voxpupuli module has since been stood up.  Anyway, I implemented quadlets as a defined type where you can just define the parameters as a hash that gets converted to systemd unit files.  This avoids the need to define a hard list of supported parameters and should allow the module to support future keywords to quadlet files with no module changes.  I also switched over rootless users and services to use the systemd module instead of the collection of exec resources used previously.
I'm going to close this request since I'm guessing your work is basically the voxpupuli implementation.

**Pull request closed without merging**